rholang-lib/src/sem/elaborator/mod.rs

```diff
@@ -0,0 +1,333 @@
+//! Elaborator performs semantic checks that go beyond basic scope resolution.
+//! It **requires** that `ResolverPass` has already run to build scopes and resolve
+//! The elaborator is stateless and can be reused for multiple for-comprehensions.
+//! ## Usage
+//!
+//! ```ignore
+//! let mut db = SemanticDb::new();
+//! let pid = db.build_index(&ast);
+//!
+//! // Step 1: Run ResolverPass (REQUIRED)
+//! let resolver = ResolverPass::new(pid);
+//! resolver.run(&mut db);
+//!
+//! // Step 2: Run elaborator
+//! let elaborator = ForCompElaborator::new(&mut db);
+//! elaborator.elaborate_and_finalize(pid)?;
+//! ```
+
+use crate::sem::{Diagnostic, ErrorKind, PID, SemanticDb};
+use rholang_parser::ast::Proc;
+
+pub mod arrow_validator;
+
+pub struct ForCompElaborator<'a, 'ast> {
+    db: &'a mut SemanticDb<'ast>,
+    diagnostics: Vec<Diagnostic>,
+}
+
+impl<'a, 'ast> ForCompElaborator<'a, 'ast> {
+    pub fn new(db: &'a mut SemanticDb<'ast>) -> Self {
+        Self {
+            db,
+            diagnostics: Vec::new(),
+        }
+    }
+
+    /// Elaborate a for-comprehension
+    ///
+    /// Prerequisites: ResolverPass MUST have run
+    pub fn elaborate_and_finalize(mut self, pid: PID) -> Result<(), Vec<Diagnostic>> {
+        if let Some(diagnostic) = self.verify_for_comprehension(pid) {
+            self.diagnostics.push(diagnostic);
+            return self.finalize();
+        }
+
+        if let Some(diagnostic) = self.validate_arrow_types(pid) {
+            self.diagnostics.push(diagnostic);
+        }
+
+        self.finalize()
+    }
+
+    /// Verify that PID is a for-comprehension
+    fn verify_for_comprehension(&self, pid: PID) -> Option<Diagnostic> {
+        let proc_ref = match self.db.get(pid) {
+            Some(p) => p,
+            None => return Some(Diagnostic::error(pid, ErrorKind::InvalidPid, None)),
```

Comment:
As I said earlier - this seems to be "your" error (passing invalid PID), not users'? Would you agree?